Added "outputLogs" Configuration Option

ObliviousHarmony opened the pull request:
### Description

This pull request builds on #822 by adding an `"outputLogs"` option to `turbo.json`. We've found that the default behavior is very noisy and doesn't provide any value to us. Rather than appending `--output-logs` every time we use a command, it would be nice to set the configuration to do this by default.

* Defaults to "full" in order to maintain backward compatibility with the current behavior.
* `--output-logs` overrides the configuration option.
* Merged `cacheHitLogsMode` and `cacheMissLogsMode` into a single type. Given the logic around this behavior, it seemed unnecessary to maintain two separate internal options when they provided no clear benefit.
* Refactored `LogsMode` into the `util` package to avoid circular dependencies in `run`, `runcache`, and `fs`.
  * I went with the `TaskOutputMode` naming because it feels like we're talking more about the task output than the log output. As far as the end-user is concerned, whether it's cached or not, this is the output of the `package.json` task.
* Added `"outputLogs": "full"` to root `turbo.json` for clarity.
* Documented the new configuration option and override behavior for `--output-logs`.

### Questions

* Should we name this `"taskOutputMode"` for consistency and a better description of what it does?
* Should this be a task configuration rather than a top-level configuration? It would be a trivial change to make in this PR and would support having a configuration where a `"build"` task has no output but another does. I'm not sure if there's any use to that though?

### Testing Instructions

1. Check `./turbow.sh run build`, behavior should match the current
2. Change `"outputLogs"` in `turbo.json` to `"none"`
3. Check `./turbow.sh run build`, behavior should match current with `--output-logs=none`
4. Change `"outputLogs"` in `turbo.json` to `"hash-only"`
5. Check `./turbow.sh run build`, behavior should match current with `--output-logs=hash-only`
4. Change `"outputLogs"` in `turbo.json` to `"new-only"`
5. Check `./turbow.sh run build`, behavior should match current with `--output-logs=new-only`
6. Check `./turbow.sh run build --output-logs=none`, behavior should match current with `--output-logs=none`
7. Optionally check other `--output-logs` inputs and verify that they override the configuration and match the current

ObliviousHarmony commented:
As an aside, this is my first contribution to a project written in Go. While I've played around with it in the past, I have never really done anything meaningful with it. Whether this pull request moves forward or not, please let me know where I've gone wrong and what I should improve next time!

ObliviousHarmony commented on cli/internal/fs/turbo_json_test.go:
```diff
@@ -76,3 +77,35 @@ func Test_ReadTurboJSON(t *testing.T) {
 	assert.EqualValues(t, remoteCacheOptionsExpected, turboJSON.RemoteCacheOptions)
 }
+
+func Test_ValidateTurboJson(t *testing.T) {
```
I based this testing pattern on what seems to be relatively common in the codebase. Is this a common testing pattern in the Go ecosystem, as opposed to having many granular test functions?

ObliviousHarmony commented on cli/internal/util/task_output_mode.go:
```diff
@@ -0,0 +1,34 @@
+package util
```
I went with the `util` package for this file because the `status.go` file feels similar in nature. Does that sound right?

ObliviousHarmony replied:
I touched on this in the PR description, but, I needed to put this in another package to avoid the circular dependency in the `turbo_json.go` validation. 

From a structural perspective, it also feels better elsewhere since it's not necessarily about `runcache` output so much as it is about the display of task output. It's true all of that output currently lives in `runcache`, but, ideologically this feels distinct enough to live in another package.

ObliviousHarmony commented on cli/internal/runcache/runcache.go:
```diff
@@ -59,77 +52,57 @@ func AddFlags(opts *Opts, flags *pflag.FlagSet) {
 	}
 }
 
-var _outputModeHelp = `Set type of process output logging. Use full to show
```
I moved this into the flag definition because it didn't feel like having it as a separate variable added anything. Was that correct or is there some benefit I missed here?

gsoltis commented on cli/internal/run/run_test.go:
```diff
@@ -40,8 +40,10 @@ func TestParseConfig(t *testing.T) {
-				runcacheOpts: runcache.Opts{},
-				scopeOpts:    scope.Opts{},
+				runcacheOpts: runcache.Opts{
+					TaskOutputMode: util.FullTaskOutput,
```
In Go, whenever possible, we try to make the zero-value for a variable or field in a struct meaningful. You can see this in practice with `CacheHitLogsMode` and `CacheMissLogsMode` in [this struct](https://github.com/vercel/turborepo/blob/main/cli/internal/runcache/runcache.go#L44). This is using Go's version of an `enum`, which is defined [here](https://github.com/vercel/turborepo/blob/main/cli/internal/runcache/runcache.go#L29). `iota` maps to 0, so that when a `LogsMode` is not specified, we default to `FullLogs`. 

I think we want the same to apply to this new field. That likely means following a similar pattern of using an enum and translating the string inputs to enum values.

That will let us avoid all the changes setting a default value of `FullTaskOutput`, as that would be the zero value for the new field.

gsoltis reviewed (changes requested):
Thanks for your contribution! I think this is a good idea, and I like the idea of getting to per-task configuration as well.

As a general structural comment on the code, I *think* we can do the following:
 * Change `TaskOutputMode` to a Go-style enum, with `FullTaskOutput` as the zero value.
 * Implement `UnmarshalJSON` for `TaskOutputMode`. This is where you can do error checking that a valid string has been provided and convert from the string to the enum value. You can see an example of this pattern [here](https://github.com/vercel/turborepo/blob/main/cli/internal/fs/turbo_json.go#L134) which gets called for each task definition when we are deserializing the pipeline.
  
I think that will let us skip the validation step as well as specifying a bunch of default value.

ObliviousHarmony commented on cli/internal/util/task_output_mode.go:
```diff
@@ -0,0 +1,34 @@
+type TaskOutputMode string
+
+// FullTaskOutput will show all task output
+const FullTaskOutput TaskOutputMode = "full"
```
I already made this change in a commit while you were reviewing 😄

ObliviousHarmony commented:
Thanks for the review and the comprehensive information @gsoltis. Originally I had avoided using an enum because I didn't want to convert back and forth between representations. 

I've gone ahead and renamed "outputLogs" to "outputMode" and moved it into the pipeline configuration. I played around with it and I like that it's more expressive. I'm much happier with the PR now and it feels like some of those rougher edges are cleaned up. Thanks again for the guidance here!

ObliviousHarmony commented on cli/internal/runcache/runcache.go:
```diff
@@ -66,77 +54,59 @@ func AddFlags(opts *Opts, flags *pflag.FlagSet) {
 
-func (l *logsModeValue) Type() string {
-	return "full|none|hash-only|new-only"
+func (l *taskOutputModeValue) Type() string {
```
Honestly, this feels a bit excessive. We're already naming the strings in the description for the command, we should assume any new options added to the enum will be added here as a matter of consistency.

ObliviousHarmony replied:
I'll leave it. Even if it's missed in the description for some reason, it's better to document that the option exists rather than miss the fact that it does.

ObliviousHarmony commented on cli/internal/runcache/runcache.go:
```diff
@@ -16,48 +17,35 @@ import (
-	LogReplayer       LogReplayer
+	SkipReads              bool
+	SkipWrites             bool
+	TaskOutputModeOverride string
```
I used a string here to abuse the zero-value of the string to indicate whether or not the value was overridden. The value is only set when the flag is present, so, it's an empty string otherwise. I also considered using a `bool` to mark the override and the actual type but that seemed excessive.

gsoltis commented on cli/internal/runcache/runcache.go:
```diff
@@ -155,30 +125,54 @@ type TaskCache struct {
 }
 
+// getTaskOutputMode returns the output mode for a given task with support for overriding it using the flag
+func (tc TaskCache) getTaskOutputMode() (util.TaskOutputMode, error) {
```
Similarly, this feels like it should be a field set on `TaskCache`, rather than a method called. We have all .of the inputs when creating the `TaskCache` instance.

ObliviousHarmony commented on cli/internal/runcache/runcache.go:
```diff
@@ -16,48 +17,35 @@ import (
 	flags.BoolVar(&opts.SkipReads, "force", false, "Ignore the existing cache (to force execution).")
 	flags.BoolVar(&opts.SkipWrites, "no-cache", false, "Avoid saving task results to the cache. Useful for development/watch tasks.")
+
+	defaultTaskOutputMode, _ := util.ToTaskOutputModeString(util.FullTaskOutput)
```
This makes sense.

ObliviousHarmony commented:
Thanks, @gsoltis. I've changed the `"--output-logs"` handling to remove the override string in favor of adding the overridden enum into the `TaskCache` struct. I'm not sure I like the ambiguity of having the `TaskOutputMode` member in the `Opts` struct, but, it looks like `viper` will help deal with that in a cleaner way. With that forward-looking context, however, I agree this solution is preferable.

gsoltis commented:
@ObliviousHarmony going through this, I think I may have given you a bad suggestion. We do in fact need to know whether or not the passed-in logs mode is an override, because when we go to make the decision about the logs mode for a particular task, we need to know if there has been a global override or not, otherwise the task's configuration should take precedence. I think the consequence of having to plumb the flags throughout the execution path makes it fairly clear.

Since I sent you down the wrong path, I'm going to send a PR to your branch shortly with my proposed fix for the situation. You can accept or adapt as you see fit.

The rest of the PR looks great.

ObliviousHarmony commented:
I just saw your PR @gsoltis, that answers my question :smile: I forgot that `RunCache` ends up with the options too. I've merged your PR, made a small change, and merged from `main`.
